core/drivers/crypto/caam/mkvb/caam_mkvb.c

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019 Bryan O'Donoghue
+ * Copyright 2019 NXP
+ *
+ * Bryan O'Donoghue <bryan.odonoghue@linaro.org>
+ */
+
+#include <caam_common.h>
+#include <caam_hal_ctrl.h>
+#include <caam_jr.h>
+#include <caam_trace.h>
+#include <caam_utils_mem.h>
+#include <drivers/imx_caam_mkvb.h>
+#include <stdint.h>
+#include <tee/cache.h>
+#include <string.h>
+#include <mm/core_memprot.h>
+
+TEE_Result caam_get_mkvb(uint8_t *dest)
+{
+	struct caam_jobctx jobctx = { 0 };
+	TEE_Result ret = TEE_ERROR_SECURITY;
+	enum caam_status res = CAAM_NO_ERROR;
+	struct caambuf buf = {};
+	uint32_t *desc = NULL;
+
+	res = caam_calloc_align_buf(&buf, 32);
+	if (res != CAAM_NO_ERROR)
+		goto out_buf;
+
+	desc = caam_calloc_desc(5);
+	if (!desc)
+		goto out_buf;
+
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, MKVB_DESC_SEQ_OUT);
+	caam_desc_add_ptr(desc, buf.paddr);
+	cache_operation(TEE_CACHEFLUSH, buf.data, MKVB_SIZE);
+	caam_desc_add_word(desc, MKVB_DESC_BLOB);
+	HASH_DUMPDESC(desc);
+
+	jobctx.desc = desc;
+	ret = caam_jr_enqueue(&jobctx, NULL);
+	DMSG("caam ret: %d", ret);
+	if (ret == CAAM_JOB_STATUS) {
+		EMSG("HuK generation failed: Job status 0x%08" PRIx32,
+		     jobctx.status);
+		goto out_desc;
+	} else if (ret == CAAM_NO_ERROR) {
+		ret = TEE_SUCCESS;
+	} else {
+		goto out_desc;
+	}
+
+	cache_operation(TEE_CACHEINVALIDATE, buf.data, MKVB_SIZE);
+	DHEXDUMP(buf.data, MKVB_SIZE);
+
+	memcpy(dest, buf.data, MKVB_SIZE);
+	ret = TEE_SUCCESS;
+out_desc:
+	caam_free_desc(&desc);
+out_buf:
+	caam_free_buf(&buf);
+	/* Increment PRIBLOB to 11 to lock out same MKVB from normal world */
+	caam_hal_ctrl_inc_priblob();
```

Comment:
Correct, I'd like to make this configurable than. We can argue about the default, I fear that users forget to set the PRIBLOB bits in the bootloader.

Reply:
> Other point, the PRIBLOB is only used in Trusted mode execution, meaning that when same descriptor is executed in normal world, the master key derived is not the same regardless the PRIBLOB settings.

I haven't found more information on this in the SRM. Is this also true for the CAAM contained in older i.MX6 SoCs like i.MX6Q? The SRM does not mention a jobring owner there and at least with HAB the TZ can not be set.
My understanding of the SRM is that the CAAMs Trusted State has more to do with HAB than with the TrustZone accesses, if you have more information, please share.

Reply:
Have a look chapter Blob types differentiated by security state in the SRM.

Reply:
The way I read it is that this non-secure state describes a not HAB booted device and not a processor or memory access from the TrustZone non-secure world. This IMO confirmed by the state being derived from the SNVS security state, which is documented to indicate the HAB status of the SoC and not the TrustZone non-secure state. Furthermore the SRM describes a transition from trusted state to the non-secure state only through the fail state, with no way to get back to trusted state after that.

Reply:
More explanation about the PRIBLOB. On i.MX6Q (and some other legacy devices), the PRIBLOB sets to 0x3 is a workaround because the Trustzone bit is not connected to the CAAM secure state machine. The problem of setting these value is disabling the CAAM BLOB interest in OPTEE because the same level of protection is used regardless the world running.
Hence, setting the PRIBLOB to 0x3 must remain a **platform specific workaround** done before switching to normal OS boot. It can not be part of CAAM driver at all.

I advice you to add a specific function into plat-imx files that is enabled or not with a CFG_XXX device to let other user to enable it if wanted. Then call this function when the OPTEE boot ends.
Does it looks reasonable for you?